src/internal/connector/graph/middleware.go

```diff
@@ -190,9 +225,14 @@ func (middleware RetryHandler) retryRequest(
 	executionCount int,
 	cumulativeDelay time.Duration,
 	exponentialBackoff *backoff.ExponentialBackOff,
-	respErr error,
+	initialErr error,
 ) (*http.Response, error) {
-	if (respErr != nil || middleware.isRetriableErrorCode(req, resp.StatusCode)) &&
+	ctx = clues.Add(
+		ctx,
+		"retry_count", executionCount,
+		"prev_resp_status", resp.Status)
+
+	if (initialErr != nil || middleware.isRetriableRespCode(req, resp.StatusCode)) &&
```

Comment:
This set of conditions has some cognitive load. Would be great to have a comment describing it. 
Also, dumb question: Wouldn't the errors listed in `isRetriableRespCode` cause an err too? Wondering if `initialErr` is sufficient?

Reply:
I don't know what errors you're referring to here.

Reply:
sorry, i meant errors listed in `retryableRespCodes`

Reply:
Ah, so that's a common misconception.  The standard golang http client does not return errors for non-2xx status codes.  A 400 resp code, for example, will return a valid http.Response with statusCode=400, and a nil error.  Http client errors in golang are reserved for an inability to make the request in at all (malformed config, missing necessary data, etc).

Now, in this case, we primarily use the graph api client wrapping the golang http client.  And the graph client _does_ transform non-2xx responses into errors (it also does not return the http.Response, so this transformation is somewhat necessary).  But since the middleware intercepts the response before that transformation exists, we're still working with response codes, not error interfaces.

Reply:
Thanks, this is useful!